README.md
# vite-plugin-rsw

> wasm-pack plugin for Vite

[![npm](https://img.shields.io/npm/v/vite-plugin-rsw.svg)](https://www.npmjs.com/package/vite-plugin-rsw)
[![npm downloads](https://img.shields.io/npm/dm/vite-plugin-rsw.svg)](https://npmjs.org/package/vite-plugin-rsw)
[![vite version](https://img.shields.io/badge/Vite-^2.0.0-000000?style=flat&labelColor=646cff)](https://github.com/vitejs/vite)
[![chat](https://img.shields.io/badge/chat-discord-blue?style=flat&logo=discord)](https://discord.gg/euyYWXTwmk)

[![awesome-rsw](./assets/awesome-rsw.svg)](https://github.com/lencx/awesome-rsw)
[![Rust](https://img.shields.io/badge/-Rust-DEA584?style=flat&logo=rust&logoColor=000000)](https://www.rust-lang.org)
[![WebAssembly](https://img.shields.io/badge/-WebAssembly-654FF0?style=flat&logo=webassembly&logoColor=ffffff)](https://webassembly.org)

|rsw version|vite version|
|---|---|
| >= `2.0.0`| >= `2.8.0`|
| >= `1.8.0`| >= `2.4.0`|
|`1.7.0`|`2.0.0 ~ 2.3.8`|

## Features

- `HMR`
- `Friendly error message` - browser and terminal

## Pre-installed

* [rust](https://www.rust-lang.org/learn/get-started)
* [nodejs](https://nodejs.org)
* [wasm-pack](https://github.com/rustwasm/wasm-pack)
* [rsw-rs](https://github.com/lencx/rsw-rs): `rsw = rs(rust) → w(wasm)` - A command-line tool for automatically rebuilding local changes, based on the wasm-pack implementation.

## Usage

### Step 1

Install rsw

```bash
cargo install rsw
```

Install vite-plugin-rsw

```bash
# With NPM:
npm i -D vite-plugin-rsw

# With Yarn:
yarn add -D vite-plugin-rsw
```

### Step 2

Edit vite.config.ts

```js
import { defineConfig } from 'vite';
import ViteRsw from 'vite-plugin-rsw';

export default defineConfig({
  plugins: [
    ViteRsw(),
  ],
})
```

### Step 3

Edit package.json

```json5
"scripts": {
    "dev": "vite",
+   "rsw": "rsw",
+   "build": "rsw build && tsc && vite build",
}
```

### Step 4

[rsw.toml options](https://github.com/rwasm/rsw-rs#options)

1. Initial rsw.toml

   ```bash
   # yarn rsw -h
   yarn rsw init
   ```

2. Generate rust crate

   ```bash
   # rsw.toml
   [new]
   # using: wasm-pack | rsw | user, default is `wasm-pack`
   using = "wasm-pack"
   ```

   ```bash
   yarn rsw new rsw-hello
   ```

3. Edit rsw.toml

   ```toml
   # link type: npm | yarn | pnpm, default is `npm`
   cli = "npm"

   [[crates]]
   name = "rsw-hello"
   # <npm|yarn|pnpm> link
   # ⚠️ Note: must be set to `true`, default is `false`
   link = true
   ```

### Step 5

1. Run rsw watch

   ```bash
   # rust crates - automatic build
   # ⚠️ Note: Do not exit the process after the command has started.
   yarn rsw watch
   ```

2. Run dev

   ```bash
   # Front-end: start dev server
   yarn dev
   ```

⚠️ Note: Open two terminal windows, execute `yarn rsw watch` in the first and `yarn dev` in the second. **The order of execution is important, do not close the first window!**

### Step 6

Deploy

```bash
yarn build
```

### Example

* [Demo](https://github.com/lencx/learn-wasm) - 🎲 Learning WebAssembly
* [Oh My Box](https://github.com/lencx/OhMyBox) - 🔮 Development toolbox, and more...

## Related

[create-mpl](https://github.com/lencx/create-mpl) - ⚡️ Create a project in seconds!

```bash
# Quickly initialize a wasm project

# npm 6.x
npm init mpl@latest my-app --type wasm

# npm 7+, extra double-dash is needed:
npm init mpl@latest my-app -- --type wasm
```

---

## 微信

> 群二维码已过期，关注公众号《浮之静》，发送“进群”，我将拉你进群一起学习。

<img height="180" src="./assets/wasm-qrcode.png" alt="wasm-wechat-qrcode" /> <img height="160" src="./assets/fzj-qrcode.png" alt="fzj-qrcode" />

## License

MIT License © 2021 [lencx](https://github.com/lencx)


CHANGELOG.md
## 2.0.8 (2022-06-27)

- fix: reload
## 2.0.0 (2022-02-27)

- a major upgrade based on `rsw-rs`

## 1.10.0 (2021-10-02)

### feature

- [Add plugin options](https://github.com/lencx/vite-plugin-rsw/issues/22)

## 1.9.1 (2021-08-24)

### fix

- [Whitespaces in project path do not work](https://github.com/lencx/vite-plugin-rsw/issues/20)

## 1.9.0 (2021-08-22)

- [Watch for dependent workspace crates](https://github.com/lencx/vite-plugin-rsw/issues/18)
- [Automatically generate .rsw.json](https://github.com/lencx/vite-plugin-rsw/issues/19)

## 1.8.0 (2021-07-13)

- remove replace: Vite v2.4.x support new URL(url, import.meta.url)

## 1.7.0 (2021-07-09)

- [x] rsw options
  - cli
  - root
  - unLinks
  - crates
- [x] startup optimization
- [x] enable debug mode
- [x] friendly error message
- [x] automatically generate template
- [x] multiple rust crate


src/types.ts
import type { ViteDevServer } from 'vite';

export type NpmCmdType = 'install' | 'link' | 'unlink';

export type CliType = 'npm' | 'pnpm';

export type TargetType = 'bundler' | 'web' | 'nodejs' | 'no-modules';

export type ProfileType = 'dev' | 'release' | 'profiling';

// feat: https://github.com/lencx/vite-plugin-rsw/issues/22
export type RswCrateOptions = {
  // support `--out-name` and `scope`
  name: string;
  // `--out-dir`
  outDir?: string;
  // defalut: 'web'
  target?: TargetType;
  // default: 'dev'
  profile?: ProfileType;
  // no default value
  mode?: 'no-install' | 'normal';
  // extra options
  extraOpts?: string[];
  // https://github.com/lencx/vite-plugin-rsw/issues/24
  // when crate is built in a watched directory,
  // stop watching files, directories, or glob patterns, takes an array of strings.
  unwatch?: string[];
}

export type WasmFileInfo = {
  fileName: string;
  source: string | false | undefined | Uint8Array;
}

export type BaseRswConfig = {
  root?: string; // default: project root
  // feat: https://github.com/lencx/vite-plugin-rsw/issues/14
  cli?: CliType; // default: npm
  unLinks?: Array<string|RswCrateOptions>;
  profile?: ProfileType;
  target?: TargetType;
}

// Plugin options
export type RswPluginOptions = BaseRswConfig & {
  crates: Array<string|RswCrateOptions>;
  // stop watching files, directories, or glob patterns, takes an array of strings
  unwatch?: string[];
}

export type CompileOneOptions = {
  config: BaseRswConfig;
  crate: string | RswCrateOptions;
  sync: boolean;
  serve?: ViteDevServer;
  filePath?: string;
  root?: string;
  outDir?: string;
}

export type RswCompileOptions = {
  config: RswPluginOptions;
  root: string;
  crate?: string | RswCrateOptions;
  serve?: ViteDevServer;
  filePath?: string;
  npmType?: NpmCmdType;
  cratePathMap?: Map<string, string>;
}

export type watchOptions = {
  paths: string[];
  unwatch: string[];
  type: 'repo' | 'crate' | 'deps';
  callback: (path: string) => void;
}
